File: Algo/Leetcode/739DailyTemperatures.cpp
// 单调栈
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> res(n, 0);
        stack<int> s;
        for(int i=0; i<n; i++){
            while(!s.empty() && temperatures[s.top()] < temperatures[i]){
                res[s.top()] = i-s.top();
                s.pop();
            }
            s.push(i);
        }
        return res;
    }
};

File: Algo/Leetcode/739DailyTemperatures.py
class Solution(object):
    def dailyTemperatures(self, temperatures):
        """
        :type temperatures: List[int]
        :rtype: List[int]
        """
        d = {}
        for i, n in enumerate(temperatures):
            d[i] = i

        st = []
        for i, n in enumerate(temperatures):
            while len(st) > 0 and st[-1][1] < n:
                top = st.pop()
                d[top[0]] = i
            st.append((i, n))

        res = []
        for i, n in enumerate(temperatures):
            res.append(d[i] - i)

        return res